src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -96,6 +98,76 @@ int distance(BoundDecisionDagNode x)
             return result.ToImmutableAndFree();
         }
 
+        /// <summary>
+        /// Enumerates the paths from the root node to the node of interest, and invokes the handler
+        /// on each one until the handler returns false.
+        /// The order is deterministic, but we're not starting from the shortest path.
+        /// </summary>
+        /// <param name="rootNode">The root node of the DAG.</param>
+        /// <param name="targetNode">The node of interest.</param>
+        /// <param name="nullPaths">Whether to permit following paths that test for null.</param>
+        /// <param name="handler">Handler to call back for every path to the target node.</param>
+        private static void VisitPathsToNode(BoundDecisionDagNode rootNode, BoundDecisionDagNode targetNode, bool nullPaths,
+            Func<ImmutableArray<BoundDecisionDagNode>, bool, bool> handler)
+        {
+            var pathBuilder = new ArrayBuilder<BoundDecisionDagNode>();
+            exploreToNode(rootNode, currentRequiresFalseWhenClause: false);
+            return;
+
+            // Recursive exploration helper
+            // Returns true to continue, false to stop
+            bool exploreToNode(BoundDecisionDagNode currentNode, bool currentRequiresFalseWhenClause)
+            {
+                if (currentNode == targetNode)
+                {
+                    return handler(pathBuilder.ToImmutable(), currentRequiresFalseWhenClause);
+                }
+
+                pathBuilder.Push(currentNode);
+
+                switch (currentNode)
+                {
+                    case null:
+                    case BoundLeafDecisionDagNode:
+                        break;
+
+                    case BoundTestDecisionDagNode test:
+                        bool skipWhenTrue = test.Test is BoundDagExplicitNullTest && !nullPaths;
+                        bool skipWhenFalse = test.Test is BoundDagNonNullTest && !nullPaths;
+
+                        if (!skipWhenTrue)
+                        {
+                            if (!exploreToNode(test.WhenTrue, currentRequiresFalseWhenClause))
+                                return false;
+                        }
+
+                        if (!skipWhenFalse)
+                        {
+                            if (!exploreToNode(test.WhenFalse, currentRequiresFalseWhenClause))
+                                return false;
+                        }
+
+                        break;
+
+                    case BoundEvaluationDecisionDagNode evaluation:
+                        if (!exploreToNode(evaluation.Next, currentRequiresFalseWhenClause))
+                            return false;
+
+                        break;
+
+                    case BoundWhenDecisionDagNode whenNode:
+                        pathBuilder.Pop();
```

Comment:
>pathBuilder.Pop(); [](http://example.com/codeflow?start=24&length=18)

Why `Pop()` the `BoundWhenDecisionDagNode`?

Reply:
That's the same logic as `ShortestPathToNode`. I think the reason is simply that `when` clauses don't contribute to narrowing down to an example.